csrc/preseg_passes/move_pad.cpp

```diff
@@ -0,0 +1,530 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <preseg_passes/move_pad.h>
+
+#include <expr_simplifier.h>
+#include <fusion.h>
+#include <ir/builder.h>
+#include <ir/interface_nodes.h>
+#include <ir/internal_base_nodes.h>
+#include <ops/alias.h>
+#include <ops/arith.h>
+#include <ops/utils.h>
+#include <transform_replay.h>
+
+namespace nvfuser::preseg_passes {
+
+namespace {
+
+// Simple PadOp is defined as a pad op with:
+// 1. zero pad value; and
+// 2. non-negative pad widths.
+bool isSimplePadOp(PadOp* pad) {
+  if (!simplifyExpr(pad->value())->isZero() &&
+      !simplifyExpr(pad->value())->isFalse()) {
+    return false;
+  }
+  // TODO: we cannot seem to always been able to prove that pad widths are >= 0.
+  // But we know that PadOp + CatOp implies such condition. Adding this
+  // short-cut to ensure that we recognized such pattern.
+  if (pad->out()->uses().size() == 1 && pad->out()->uses()[0]->isA<CatOp>()) {
+    return true;
+  }
+  for (Val* pad_val : pad->getPadWidths()) {
+    if (!simplifyExpr(
+             SimplifyingIrBuilder::geExpr(pad_val, pad->fusion()->zeroVal()))
+             ->isTrue()) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// check if `use` and `p` are the same pad operatoin.
+bool isSamePadOp(Expr* use, PadOp* p) {
```

Comment:
Did you try https://github.com/NVIDIA/Fuser/blob/9f8c99148c2299ff3a22f2e8f2d9a86bc4063355/csrc/ir/base_nodes.cpp#L344?

Reply:
I don't think that one would work, since it doesn't try to simplify pad widths, but rather rely on things coming from the same definition https://github.com/NVIDIA/Fuser/blob/3233bcef198f2693634a3e1b49a9836bf4d28694/csrc/ir/base_nodes.cpp#L147-L149

Reply:
This seems useful so in another PR it could probably be generalized into an option like `use->sameAs(p, /*shallow=*/false)` or something.

Reply:
That sounds reasonable.

I guess we can just iterate through all inputs and check for `!simplifyExpr(SimplifyingIrBuilder::eqExpr(a->input(i), b->input(i)))` for non-TensorView inputs?

Conceptually it felt a bit blurry to me. But I see how it could be useful.